File: dart-sdk/lib/src/attributes.dart
import 'package:meta/meta.dart';

import 'package:eppo_sdk/src/rust/api/attributes.dart' as core;
import 'package:eppo_sdk/src/rust/api/client.dart' as core;

enum AttributeKind {
  categorical,
  numeric,
}

final class Attributes {
  @internal
  final Map<String, core.AttributeValue> coreAttributes = new Map();

  Attributes();

  void stringAttribute(String key, String value) {
    this.coreAttributes[key] = core.stringAttribute(value);
  }

  void numberAttribute(String key, double value, {AttributeKind kind = AttributeKind.numeric}) {
    this.coreAttributes[key] = kind == AttributeKind.numeric ? core.numericAttribute(value) : core.categoricalNumberAttribute(value);
  }

  void boolAttribute(String key, bool value) {
    this.coreAttributes[key] = core.boolAttribute(value);
  }
}

File: dart-sdk/rust/src/api/attributes.rs
use flutter_rust_bridge::frb;

pub use eppo_core::AttributeValue;

#[frb(external)]
impl AttributeValue {}

#[frb(sync, positional)]
pub fn numeric_attribute(value: f64) -> AttributeValue {
    AttributeValue::numeric(value)
}

#[frb(sync, positional)]
pub fn categorical_number_attribute(value: f64) -> AttributeValue {
    AttributeValue::categorical(value)
}

#[frb(sync, positional)]
pub fn string_attribute(value: String) -> AttributeValue {
    AttributeValue::categorical(value)
}

#[frb(sync, positional)]
pub fn bool_attribute(value: bool) -> AttributeValue {
    AttributeValue::categorical(value)
}
